### ISSUE TITLE
[v0.9.3] terse error - panic: could not convert row to (T1, T2)

### ISSUE THREAD

**Issue by boxofrox** [blackbeam/mysql_async#10]
It took me more than a few hours track this bug down, and even then I'm not sure I found the cause.  This issue is mostly to document/discuss how I troubleshot this panic, and whether we can patch the code to improve the errors, or provide a notice in the docs/comments.

I'm running this bit of code against a query with a large result set (approx 109,000 rows)
```rust
pool.get_conn()
    .and_then(move |conn| conn.query(&sql))
    .and_then(|result| {
        result.map(|row|{
            let (id, text) : (String, String) = mysql_async::from_row(row);  // panic here
            Data::new(id, text)
        })
    })
    //...
```

The table consists of an primary `integer(11)` field, `id`, and a `varchar(255)` field, `text`.

The backtrace for this error did little to help me identify the cause, and I wasted time trying to figure out why `src/libcore/option.rs` was panicking.

```
thread '<unnamed>' panicked at 'Could not convert row to (T1,T2)', /checkout/src/libcore/option.rs:823
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:355
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:371
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:549
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:511
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:495
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:471
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:69
   9: core::option::expect_failed
             at /checkout/src/libcore/option.rs:823
  10: <core::option::Option<T>>::expect
             at /checkout/src/libcore/option.rs:302
  11: <(T1, T2) as mysql_async::value::FromRow>::from_row
             at /home/charetjc/files/development/rust/crates/mysql_async/src/value.rs:380
  12: mysql_async::value::from_row
             at /home/charetjc/files/development/rust/crates/mysql_async/src/value.rs:280
  13: fuzzyd::database::fetch_choices::{{closure}}::{{closure}}
             at src/bin/fuzzyd/database.rs:101
----snipped----
```

Eventually I found `Option` is panicking because of code in `mysql_async/src/value.rs` like [line 380](https://github.com/blackbeam/mysql_async/blob/v0.9.3/src/value.rs#L380)

```rust
        FromRow::from_row_opt(row).ok().expect("Could not convert row to (T1,T2)")
```

While `from_row_opt` returns a `Result`, `.ok()` converts the `Result` into an `Option` that discards the `Err`, so that `.expect()` doesn't print the `Err`??

I can understand this somewhat from a security perspective, since the `Err` contains potentially sensitive data that shouldn't find its way into logs.  Also, the `Err` is pretty complex and will have a duplicate stack trace (if `RUST_BACKTRACE=1`), which can be confusing due to the poor formatting of `Debug` output.

That said, I forked `v0.9.3` and dropped the call to `.ok()`, found...

```
thread '<unnamed>' panicked at 'Could not convert row to (T1,T2):

Error
  ( FromRow
    ( Row
      { values: [ Some( Bytes( [ 85, 80, 87, 45, 53, 55, 45, 49, 48, 48, 48, 45, 50 ] ) )
                , Some( NULL )
                ]
      , columns:
        [ Column
          { payload: [REDACTED]
          , schema: (5, 14)
          , table: (20, 9)
          , org_table: (30, 9)
          , name: (40, 2)
          , org_name: (43, 3)
          , default_values: None
          , column_length: 96
          , character_set: 45
          , flags: NOT_NULL_FLAG | MULTIPLE_KEY_FLAG | PART_KEY_FLAG
          , column_type: MYSQL_TYPE_VAR_STRING
          , decimals: 0
          }
        , Column
          { payload: [REDACTED]
          , schema: (5, 14)
          , table: (20, 9)
          , org_table: (30, 9)
          , name: (40, 8)
          , org_name: (49, 17)
          , default_values: None
          , column_length: 1020
          , character_set: 45
          , flags: 
          , column_type: MYSQL_TYPE_VAR_STRING
          , decimals: 0
          }
        ]
      }
    )
  , State
    { next_error: None
    , backtrace: None
    }
  )

', /checkout/src/libcore/result.rs:859
```

...and with this, I can see that one of my two values is `NULL`, which is a problem when my code expects a `String`.

From here, I'm not certain where to go.  I will try using `let (id, text) : (String, Option<String>) = from_row()` and see if that works, but I didn't see anything in the docs that indicated this will work or how to handle `NULL` values in a SQL field.

So, with that said, what to do?  I present options!

1. Can we drop the `.ok()` from `src/value.rs`?
    a.  or can we map the error into a terse, but more informative, "Cannot convert NULL SQL value to String"? (not sure this is even possible)
2. Can we document handling `NULL` values, if so where?
3. Can we document that "Could not convert..." panics can be caused by type mismatches, if so where?

I'm happy to take a stab at PR's for these.  Also, I know `v0.10.0` is out, and I'm fairly certain I would have the same issue there.  `src/value.rs` didn't change much.  I hope to update to `v0.10.0` soon...ish, but I lost a bit of time on my project trying to offload the database routines into a separate thread, and racing to catch up.

Cheers!

**Comment by boxofrox** [comment]
I am pleased to confirm that `let (id, text) : (String, Option<String>) = from_row()` does work!

**Comment by blackbeam** [comment]
This is known problem (blackbeam/rust-mysql-simple#45), and the only solution that really helps me is to strictly follow my schema (i.e. if column is nullable then corresponding type should always be `Option<T>`).

> Can we drop the `.ok()` from `src/value.rs`?

I don't think that dumping raw row will really help here. I think it's better to alter error message to better point to `FromRow` trait (for example "FromRow convertion error: couldn't convert mysql row to requested type (T1, T2)" or even "Couldn't convert mysql row to requested type (T1, T2). (see FromRow docs)") and then alter `FromRow` documentation to point possible conversion problems (i.e. non option type for nullable column or simple type mismatch) and add an advice to use `RUST_BACKTRACE=1` to find the problematic call (i think it is `src/bin/fuzzyd/database.rs:101` in your case).

I think that situation in rust-mysql-simple is better because the `Value` documentation mentioning a way to handle nullable value, but unfortunately this documentation is somehow missed during copy to this crate.

Anyway i will alter error and docs when i find spare time (hard task since my English is awful) because this issue is annoying.

**Comment by boxofrox** [comment]
I agree that following your schema is the best.

Part of the problem on my end was 1) I failed to realize the table I was working with might (and did) contain `NULL`, and 2) I had no idea why `from_row` failed, I only knew that it did fail.  It's that second point that I want to address most.  I think converting the `Err(FromRow)` into `Err("found SQL row with (Bytes, NULL)")` and letting the `.expect()` present that addition message would be sufficient.

I'm not certain how `mysql_async` behaves in this case, but if I try to convert a SQL row with 3 fields into `(T1, T2)`, would that work and ignore the third field, or fail with `could not convert row to (T1, T2)`?  If the latter, then `"found sql row with (Bytes, Bytes, Bytes)" in the error message tells me exactly where I went wrong in my code.

Yep, `src/bin/fuzzyd/database.rs:101` was this line in my example above,

```rust
let (id, text) : (String, String) = mysql_async::from_row(row);  // panic here
```

I am interested in creating a PR, though I'm also pressed for time.  Maybe we'll see which of us gets to it first.  I'm certainly happy and willing to proofread for you.  If you can show me where you want the documentation and/or the key ideas to cover, I can also write that out.  Though, I have to say, your responses are well-written English.  Far better than what I would call awful :wink:

I wasn't aware that `blackbeam/rust-mysql-simple` was related, so I'll definitely start checking the docs there.

**Comment by blackbeam** [comment]
Fixed in v0.10.1
